python/Ice/async/client.py

```diff
@@ -3,55 +3,82 @@
+        if len(sys.argv) > 1:
+            print(sys.argv[0] + ": too many arguments")
+            return 1
 
-    c = None
-    while c != 'x':
         try:
+            hello = Demo.HelloPrx.checkedCast(communicator.propertyToProxy('Hello.Proxy'))
+        except Ice.Exception as ex:
+            print("invalid proxy:", ex)
+            return 1
+
+        menu()
+
+        tasks = set()
+        done = set()
+
+        while True:
             sys.stdout.write("==> ")
             sys.stdout.flush()
-            c = sys.stdin.readline().strip()
-            if c == 'i':
-                hello.sayHello(0)
-            elif c == 'd':
-                cb = Callback()
-                hello.sayHelloAsync(5000).add_done_callback(cb.response)
-            elif c == 's':
-                hello.shutdown()
-            elif c == 'x':
-                pass  # Nothing to do
-            elif c == '?':
-                menu()
-            else:
-                print("unknown command `" + c + "'")
-                menu()
-        except Ice.Exception as ex:
-            print(ex)
+            nextCharTask = asyncio.create_task(getNextChar())
+            tasks.add(nextCharTask)
+            while nextCharTask not in done:
+                done, tasks = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+                if nextCharTask in done:
+                    c = nextCharTask.result()
+                    if c == 'i':
+                        # create asyncio future from Ice future
+                        tasks.add(Ice.wrap_future(hello.sayHelloAsync(0)))
+                    elif c == 'd':
+                        # create asyncio task from coroutine returned by slowSayHello
+                        tasks.add(asyncio.create_task(slowSayHello(hello)))
+                    elif c == 's':
+                        tasks.add(Ice.wrap_future(hello.shutdownAsync()))
+                    elif c == 'x':
+                        # cancel all outstanding tasks
+                        for t in tasks:
+                            t.cancel()
+                        return 0
+                    elif c == '?':
+                        menu()
+                    else:
+                        print(f"unknown command {c!r}")
+                        menu()
+                else:
+                    # retrieve exceptions of done tasks, if any
+                    for t in done:
+                        try:
+                            t.result()
+                        except Ice.Exception as ex:
+                            print("task exception:", ex)
+                        except RuntimeError as ex:
+                            print("task runtime error:", ex)
+
+
+async def getNextChar():
+    # run blocking IO with default executor
+    c = await asyncio.get_running_loop().run_in_executor(None, sys.stdin.readline)
+    return c.strip()
+
+
+async def slowSayHello(hello):
+    await Ice.wrap_future(hello.sayHelloAsync(5000))
```

Comment:
Why don't we catch the exception directly here instead of having a "done" set for done tasks?

Reply:
Because this done processing works for all tasks. For example, without it, if you try to shutdown a server that is already shutdown, you'd get:
```
==> s
Future exception was never retrieved
future: <Future finished exception=exception ::Ice::ConnectionRefusedException
{
    error = 61
}>
Ice.ConnectionRefusedException: Ice.ConnectionRefusedException:
```

Reply:
However, the `else:` was not correct: we should do this processing all the time.